oak_containers_orchestrator/src/crypto.rs

```diff
@@ -0,0 +1,96 @@
+//
+// Copyright 2023 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::proto::oak::containers::v1::{
+    orchestrator_crypto_server::OrchestratorCrypto, DeriveSessionKeysRequest,
+    DeriveSessionKeysResponse, KeyOrigin,
+};
+use oak_crypto::encryptor::{EncryptionKeyProvider, RecipientContextGenerator};
+use std::sync::Arc;
+use tonic::{Request, Response};
+
+pub struct KeyStore {
+    // TODO(#4507): Remove `Arc` once the key is no longer required by the `Attester`.
+    instance_encryption_key: Arc<EncryptionKeyProvider>,
+    group_encryption_key: Arc<EncryptionKeyProvider>,
+}
+
+impl Default for KeyStore {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl KeyStore {
+    pub fn new() -> Self {
+        let instance_encryption_key = Arc::new(EncryptionKeyProvider::generate());
+        let group_encryption_key = instance_encryption_key.clone();
+        Self {
+            instance_encryption_key,
+            group_encryption_key,
+        }
+    }
+
+    pub fn instance_encryption_key(&self) -> Arc<EncryptionKeyProvider> {
+        // TODO(#4442): Currently we have to give the encryption key provider to the `ipc_server`.
+        // Once we move all enclave apps to the new crypto service and update the `Attester` to not
+        // have the private key - this function should be removed.
+        self.instance_encryption_key.clone()
+    }
+
+    pub fn instance_encryption_public_key(&self) -> Vec<u8> {
+        self.instance_encryption_key.get_serialized_public_key()
+    }
+}
+
+// TODO(#4442): Create CryptoService after group key was provisioned.
+pub(crate) struct CryptoService {
+    key_store: Arc<KeyStore>,
+}
+
+impl CryptoService {
+    pub(crate) fn new(key_store: Arc<KeyStore>) -> Self {
+        Self { key_store }
+    }
+}
+
+#[tonic::async_trait]
+impl OrchestratorCrypto for CryptoService {
+    async fn derive_session_keys(
+        &self,
+        request: Request<DeriveSessionKeysRequest>,
+    ) -> Result<Response<DeriveSessionKeysResponse>, tonic::Status> {
+        let request = request.into_inner();
+
+        let encryption_key = match request.key_origin() {
+            KeyOrigin::Unspecified => {
```

Comment:
Is there some magic conversion between `KEY_ORIGIN_UNSPECIFIED` and `Unspecified` here?

Reply:
`cargo` told me that `KeyOrigin::Unspecified` was not defined, and when I tried using the full name, it said that there is no `KeyOrigin::KeyOriginUnspecified`

Reply:
So where did `KEY_ORIGIN_UNSPECIFIED` go? It sounds like the proto compiler didn't rebuild correctly. Anyways if CI passes I guess that's fine (though I still don't understand how)